### PR TITLE
Add `log_step_metadata` utility function

### PR DESCRIPTION
**Pull request by strickvl**
Adds a `log_step_metadata` utility function that can be used from within a step as well as outside it.

See the tests for examples of it in use both within a step and outside. This also includes a brief docs update addressing the added functionality.

## Summary by CodeRabbit

- **New Features**
  - Introduced a method for logging metadata associated with pipeline steps.
- **Documentation Updates**
  - Added a new section on how to use the `log_step_metadata` method in the user guide.
- **Bug Fixes**
  - Removed the outdated `log_model_version_metadata` function to streamline metadata logging.
- **Tests**
  - Implemented tests to ensure reliable logging of step metadata in pipelines.

### CODE REVIEW

**Comment by strickvl** [comment]
@coderabbitai review

**Review by avishniakov — CHANGES_REQUESTED** [review]

File: src/zenml/__init__.py [hunk 1]
```diff
@@ -46,11 +46,11 @@
 from zenml.artifacts.artifact_config import ArtifactConfig
 from zenml.artifacts.external_artifact import ExternalArtifact
 from zenml.model.model_version import ModelVersion
-from zenml.model.utils import log_model_version_metadata
```
[the remark this answers is not in the document]
Comment (strickvl) [reply src/zenml/__init__.py:49]: Ah right. I saw it wasn't below so removed unthinkingly. I've added it back above and below now.

File: docs/book/user-guide/advanced-guide/data-management/logging-metadata.md [hunk 2]
```diff
@@ -76,7 +76,70 @@ def train_model(dataset: pd.DataFrame) -> Annotated[ClassifierMixin, ArtifactCon
...
+# run some step
+
+# subsequently log the metadata for the step
+log_step_metadata(
+        metadata={
+            "some_metadata": {
+                "a_number": 3,
+            }
+        },
+    )
```
[the remark this answers is not in the document]
Comment (strickvl) [reply docs/book/user-guide/advanced-guide/data-management/logging-metadata.md:141]: Yes good catch. Updated with correct values now.

**Comment by strickvl** [comment]
@coderabbitai review

File: src/zenml/steps/utils.py [hunk 3]
```diff
@@ -403,3 +408,67 @@ def f3(condition):
...
+            the function is not called from within a step.
+    """
+    if not step_name:
+        try:
+            step_context = get_step_context()
+            step_name = step_context.step_name
+        except RuntimeError:
+            step_context = None
+    else:
+        step_context = None
```
Comment (avishniakov) [inline src/zenml/steps/utils.py:445]: Not critical, but you can do `step_context = None` before `if not step_name:` and do `pass` on the exception catch. Will be the same logic, but 1 line less 🙂

**Review by avishniakov — APPROVED** [review]

File: src/zenml/steps/utils.py (hunk 3 above)
Comment (strickvl) [reply src/zenml/steps/utils.py]: I shaved 2 lines off ;)

Comment (avishniakov) [reply src/zenml/steps/utils.py]: 👑

**Pull request merged**